[addons] change audio decoder addon interface to new C++ style

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->
Next part of rework.

There becomes also the addon interface for audio decoder addons changed
to the new C++ style.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDecoder.h:
```diff
@@ -0,0 +1,242 @@
+    /// @param[out] length              Length of file
+    /// @return                         True on success, false on failure
+    ///
+    virtual bool ReadTag(std::string file, std::string& title, std::string& artist, int& length) { return false; }
```
“const std::string& file“?

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDecoder.h:
```diff
@@ -0,0 +1,242 @@
+    /// @return                         true if successfully done, otherwise
+    ///                                 false
+    ///
+    virtual bool Init(std::string filename, unsigned int filecache,
```
“const std::string& filename“?

AlwinEsch replied:
Was gone down during take to here, becomes changed. Thanks!

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDecoder.h:
```diff
@@ -0,0 +1,242 @@
+    /// @param[in] file                 File to read tag for
+    /// @return                         Number of tracks in file
+    ///
+    virtual int TrackCount(std::string file) { return 1; }
```
“const std::string& file“?

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDecoder.h:
```diff
@@ -0,0 +1,242 @@
+      std::string intTitle;
+      std::string intArtist;
+      memset(title, 0, ADDON_STANDARD_STRING_LENGTH_SMALL);
+      memset(artist, 0, ADDON_STANDARD_STRING_LENGTH_SMALL);
```
"memset(..." vs. "artist = {..."

AlwinEsch replied:
Think this only work on a static structure where the size can easily identified.

AlwinEsch replied:
Ah, becomes now done on Kodi side.

ksooo reviewed (commented):
some comments / suggestions

AlwinEsch commented on xbmc/addons/AudioDecoder.cpp:
```diff
@@ -16,51 +16,41 @@
-  m_strExt = std::move(strExt);
-  memset(&m_struct, 0, sizeof(m_struct));
+  m_CodecName = addonInfo->Type(ADDON_AUDIODECODER)->GetValue("@name").asString();
+  m_strExt = m_CodecName + "stream";
```
Can be done, but want to have things like them as normal function calls in case it becomes bigger to have better overview.

ksooo replied:
Sorry, I don't get why this helps to keep a better overview.

ksooo replied:
But I see the negative performance impact...

AlwinEsch replied:
Thats true :-), but found another reason to leave like them. `m_CodecName` and `m_strExt` are parts on parent class.

ksooo replied:
Isn't it possible to init m_CodecName and m_strExt in the ctor of the base class (maybe by introducing a new base class ctor taking the codecname)?

But not a big deal. Leave it as is, if you like.

AlwinEsch replied:
I'm look for it.

Was a little at the test herewith distracted: https://github.com/xbmc/xbmc/pull/12338

AlwinEsch commented on xbmc/addons/AudioDecoder.cpp:
```diff
@@ -16,51 +16,41 @@
+  if (!CreateInstance(&m_struct))
+    return false;
+
+  return true;
```
Becomes squeezed :D

notspiff commented:
Didnt read code but i want to point out that audiodecs relies on a dedicated copy of the binary per instance (which means making a copy to force the linker). A lot of them are riddled by globals.

AlwinEsch commented:
I know, thats also a very critical part where I look during changes of addons. Hope to become them changed.

But want to remove this ugly hack way where a dll becomes copied to another name to use again.

notspiff commented:
That is the workaround to force the linker. I do not think there is another way.

AlwinEsch commented:
Let me look, timidity is for them the best enemy :-D. Hope you are not right ;-)

AlwinEsch commented:
Not want to have them in, before not every addon is confirmed that it work.

AlwinEsch committed:
[addons] change audio decoder addon interface to new C++ style

Next part of rework.

There becomes also the addon interface for audio decoder addons changed
to the new C++ style.

AlwinEsch commented:
Can someone look on https://github.com/notspiff/audiodecoder.timidity/pull/17

Want to remove the hack way from Kodi where the dll is copied to another name to "dlopen" again.
If an addon needs that I think he should do it himself.

Do have this way in Kodi make for me much more problems.

notspiff commented:
Sorry dude i am drunk and listening to loud metal this weekend aka on a
festival ;)

24. jun. 2017 7:00 p.m. skrev "Alwin Esch" <notifications@github.com>:

> Can someone look on notspiff/audiodecoder.timidity#17
> <https://github.com/notspiff/audiodecoder.timidity/pull/17>
>
> Want to remove the hack way from Kodi where the dll is copied to another
> name to "dlopen" again.
> If an addon needs that I think he should do it himself.
>
> Do have this way in Kodi make for me much more problems.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/12336#issuecomment-310850770>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFktYAWygHl4-owoztBmuilCr2XP8grdks5sHUCwgaJpZM4OC9us>
> .
>

FernetMenta commented:
Roskilde? enjoy :)

AlwinEsch commented:
Enjoy it and Cheers :)

notspiff commented:
Tons of rock, halden norway

24. jun. 2017 9:05 p.m. skrev "Alwin Esch" <notifications@github.com>:

> Enjoy it and Cheers :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/12336#issuecomment-310859745>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFktYCPyN8yus64d5gJoBxwbQ1ARjOGCks5sHV4FgaJpZM4OC9us>
> .
>
